### PR TITLE
feat: custom operations

### LINKED ISSUE #1095: Custom Operations
Part of Epic: #884

### PR DESCRIPTION
**Pull request by gabcoelho**
### Related Issues

Closes #1095 

### Description and Example 

Part of Epic #884

### Checklist

Please, check if these important points are met using `[x]`:

- [x] I read the [PR Guide] and followed the process outlined there for submitting this PR.
- [x] I avoided _breaking changes_ by not changing public APIs that people rely on. 
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: iOS/Sources/Beagle/Sources/ContextExpression/Tests/OperationEvaluation/CustomOperationsTests.swift [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        // When
+        evaluateSumOperation { evaluatedResults in
+            // Then
+            XCTAssertEqual(evaluatedResults, comparableResults)
+        }
+    }
+    
+    func testInvalidName() {
+        // Given /When
+        dependencies.customOperationsProvider.register(operation: .custom("sum???"), handler: { parameters -> DynamicObject in
```
Comment (botbeaglezup) [inline iOS/Sources/Beagle/Sources/ContextExpression/Tests/OperationEvaluation/CustomOperationsTests.swift:65]: <table data-meta="generated_by_ios">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/b2fba7f30eeda153ffe3b19263a01a80c356e3cd/iOS/Sources/Beagle/Sources/ContextExpression/Tests/OperationEvaluation/CustomOperationsTests.swift#L65"></span>Unused parameter “parameters” in a closure should be replaced with _.<br />
<code>unused_closure_parameter</code> <code>CustomOperationsTests.swift:65</code></td>
    </tr>
  </tbody>
</table>

File: iOS/Sources/Beagle/Sources/ContextExpression/CustomOperationProvider.swift [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    public init(dependencies: Dependencies) {
+        self.dependencies = dependencies
+    }
+    
+    public func register(operation: Operation.Name, handler: @escaping OperationHandler) {
+        guard operation.rawValue.range(of: "^[a-zA-Z0-9_]*$", options: .regularExpression) != nil else {
+            dependencies.logger.log(Log.customOperations(.invalid(name: operation.rawValue)))
+            return
+        }
```
Comment (lucasaraujo) [inline iOS/Sources/Beagle/Sources/ContextExpression/CustomOperationProvider.swift:62]: This guard allows empty name (`""`) and digits only (`"123"`) breaking spec #884 . One possible solution would be match two regex: `#"^\w+$"#` and `#"\D+"#`.
The spec doesn't mention names starting with digits. Should it be allowed?

**Review by lucasaraujo — CHANGES_REQUESTED** [review]

File: iOS/Sources/Beagle/Sources/ContextExpression/CustomOperationProvider.swift (hunk 2 above)
Comment (gabcoelho) [reply iOS/Sources/Beagle/Sources/ContextExpression/CustomOperationProvider.swift:62]: Nice catch, I forgot about that, I'll make the changes and verify if we will allow names starting with digits. Thanks!!

**Review by victorrabelozup — COMMENTED** [review]
This solution is good, but I think that the problem of the operation type(native vs custom) is being treated in the wrong place. The parcer should be able to parce correctly for the enum type.  Another solution could be the default operations use the same approach as the custom, using a dictionary instead of a enum, the widgets are done with this idea, using the same functions and dictionary to add both native and custom widgets.

**Review by lucasaraujo — DISMISSED** [review]

**Review by dantes-git — APPROVED** [review]

**Review by lucasaraujo — APPROVED** [review]

**Pull request merged**